language-service/src/services/yamlTraversal.ts

```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+import * as Parser from "../parser/jsonParser";
+import { YAMLDocument } from "../parser/yamlParser";
+import { PromiseConstructor, Thenable } from "vscode-json-languageservice";
+import { TextDocument, Position } from "vscode-languageserver-types";
+import { YamlNodeInfo } from "../yamlLanguageService";
+
+export class YAMLTraversal {
+
+    private promise: PromiseConstructor;
+
+    constructor(promiseConstructor: PromiseConstructor) {
+        this.promise = promiseConstructor || Promise;
+    }
+
+    public findNodes(document: TextDocument, yamlDocument: YAMLDocument, key: string): Thenable<YamlNodeInfo[]> {
+        if(!document){
+            this.promise.resolve([]);
+        }
+
+        const jsonDocument = yamlDocument.documents.length > 0 ? yamlDocument.documents[0] : null;
+        if(jsonDocument === null){
+            return this.promise.resolve([]);
+        }
+
+        let nodes: YamlNodeInfo[] = [];
+        jsonDocument.visit((node => {
+            const propertyNode = node as Parser.PropertyASTNode;
+            if (propertyNode.key && propertyNode.key.value === key && propertyNode.value.start !== propertyNode.value.end) {
+                nodes.push({
+                    position: document.positionAt(node.start),
+                    key: propertyNode.key.value,
+                    value: propertyNode.value.getValue()
+                });
+            }
+            return true;
+        }));
+
+        return this.promise.resolve(nodes);
+    }
+
+    public getNodeInputValues(document: TextDocument, yamlDocument: YAMLDocument, position: Position): {[key: string]: string} {
+        if(!document){
+            return null;
+        }
+
+        const offset: number = document.offsetAt(position);
+        const jsonDocument = yamlDocument.documents.length > 0 ? yamlDocument.documents[0] : null;
+        if(jsonDocument === null){
+            return null;
+        }
+
+        // get the node by position and then walk up until we find an object node with properties
+        let node = jsonDocument.getNodeFromOffset(offset);
+        while (node !== null && !(node instanceof Parser.ObjectASTNode)) {
+            node = node.parent;
+        }
+
+        if (!node) {
+            return null;
+        }
+
+        // see if this object has an inputs property
+        const propertiesArray = (node as Parser.ObjectASTNode).properties.filter(p => p.key.value === "inputs");
+        if (!propertiesArray || propertiesArray.length !== 1) {
+            return null;
```

Comment:
This should keep going up the tree in case we support task inputs that are objects in the future.

Reply:
Do you know what that would look like? How would I test this to know my code here works rather than just hoping it will work in the future?

Reply:
Once "inputs" becomes a parameter instead of a literal, you can test it using "task" or anything else higher up in the tree.

Reply:
I've renamed this to getNodePropertyValues to make it more generic (My code will call it with "inputs") and used the name property to make clear this returns property values and not objects. If we extend the language to allow tasks and we know what they will look like then this can be updated.